File: backend/src/insightflow/sanitize.py
"""DuckDB hata mesajlarından veri değerlerini temizler.

Hata mesajları veri içerebilir: `Could not convert string 'Ahmet Yılmaz' to INT32` veya
`value 184250.75 is out of range`. Self-healing döngüsünde LLM'e yalnızca hatanın biçimi gider.
Sütun adları (çift tırnaklı) korunur; LLM onları zaten şemadan biliyor ve onarım için gerekli.
"""

import re

MAX_ERROR_LENGTH = 600

_SINGLE_QUOTED = re.compile(r"'(?:[^']|'')*'")
_NUMBER = re.compile(r"(?<![\w\"])-?\d+(?:[.,]\d+)?(?![\w\"])")
_LINE_PREFIX = re.compile(r"\bLINE \d+:")
_WHITESPACE = re.compile(r"[ \t]+")


def sanitize_error(message: str) -> str:
    text = _SINGLE_QUOTED.sub("'…'", message)
    # "LINE 1:" konum bilgisini koru, diğer sayıları maskele.
    lines = []
    for line in text.splitlines():
        prefix = _LINE_PREFIX.match(line)
        head = prefix.group(0) if prefix else ""
        lines.append(head + _NUMBER.sub("#", line[len(head):]))
    text = _WHITESPACE.sub(" ", "\n".join(lines)).strip()
    return text[:MAX_ERROR_LENGTH]

File: frontend/src/lib/sanitize.ts
/**
 * DuckDB hata mesajlarından veri değerlerini temizler (backend/src/insightflow/sanitize.py ile aynı kurallar).
 * Onarım isteği sunucuya gitmeden önce tarayıcıda uygulanır; sunucu da ayrıca temizler.
 */
const MAX_ERROR_LENGTH = 600;
const SINGLE_QUOTED = /'(?:[^']|'')*'/g;
const NUMBER = /(?<![\w"])-?\d+(?:[.,]\d+)?(?![\w"])/g;
const LINE_PREFIX = /^LINE \d+:/;

export function sanitizeError(message: string): string {
  const text = message
    .replace(SINGLE_QUOTED, "'…'")
    .split(/\r?\n/)
    .map((line) => {
      const head = line.match(LINE_PREFIX)?.[0] ?? "";
      return head + line.slice(head.length).replace(NUMBER, "#");
    })
    .join("\n")
    .replace(/[ \t]+/g, " ")
    .trim();
  return text.slice(0, MAX_ERROR_LENGTH);
}
